### PR TITLE
feat(account-lib): recover signature from raw tx

### PR DESCRIPTION
**Pull request by nachofm**
STLX-12096

### CODE REVIEW

File: modules/account-lib/test/unit/coin/dot/transactionBuilder/transactionBuilderFactory.ts [hunk 1]
```diff
@@ -53,7 +53,6 @@ describe('dot Transaction Builder Factory', () => {
       builder
         .validity({ firstValid: 3933 })
         .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
-        .sender({ address: sender.address })
         .sign({ key: sender.secretKey });
```
Comment (jrafael-simtlix) [inline modules/account-lib/test/unit/coin/dot/transactionBuilder/transactionBuilderFactory.ts:56]: Remove sign also!

File: modules/account-lib/test/unit/coin/dot/transactionBuilder/transactionBuilderFactory.ts [hunk 2]
```diff
@@ -77,7 +76,6 @@ describe('dot Transaction Builder Factory', () => {
     builder
       .validity({ firstValid: 3933, maxDuration: 64 })
       .referenceBlock('0x149799bc9602cb5cf201f3425fb8d253b2d4e61fc119dcab3249f307f594754d')
-      .sender({ address: sender2.address })
       .sign({ key: sender2.secretKey });
```
Comment (jrafael-simtlix) [inline modules/account-lib/test/unit/coin/dot/transactionBuilder/transactionBuilderFactory.ts:79]: ditto

**Review by jrafael-simtlix — DISMISSED** [review]

File: modules/account-lib/src/coin/dot/transactionBuilder.ts [hunk 3]
```diff
@@ -206,6 +208,8 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
     this.transaction.chainName(this._material.chainName);
     if (this._keyPair) {
       this.transaction.sign(this._keyPair);
+    } else if (this._signature) {
+      this.transaction.addSignature(this._signature);
```
Comment (byhow) [inline modules/account-lib/src/coin/dot/transactionBuilder.ts:212]: This method is from the builder not transaction: #1990

Comment (byhow) [reply modules/account-lib/src/coin/dot/transactionBuilder.ts:212]: https://github.com/BitGo/BitGoJS/pull/1990/files#diff-b4a14cb39fe8cc16ac38f9ccb37aabb1e4b64c4dfe6596d39fd6f2e24437248eR375-R380

**Review by byhow — CHANGES_REQUESTED** [review]
There is some difference between this PR and the signing pr. Let unify them before merging this in

**Comment by byhow** [comment]
Hi @nachofm just FYI, we were implementing the TSS signature tests but it needed the recovery method from this PR, so we  
 added the method there but without the unit tests from here. PR here: #2017 & #1990

**Comment by lcovar** [comment]
Hey @nachofm , we can close this PR as we combined your work with our tss pr. 

https://github.com/BitGo/BitGoJS/pull/2017

**Comment by byhow** [comment]
@lcovar I was thinking we can keep this for the unit tests portion

File: modules/account-lib/src/coin/dot/utils.ts [hunk 4]
```diff
@@ -186,8 +199,18 @@ export class Utils implements BaseUtils {
    * @returns {string}
    */
   decodeDotAddress(address: string, ss58Format?: number): string {
-    const keypair = new KeyPair({ pub: Buffer.from(decodeAddress(address, undefined, ss58Format)).toString('hex') });
-    return keypair.getAddress();
+    return this.decodeDotAddressToKeyPair(address, ss58Format).getAddress();
```
Comment (byhow) [inline modules/account-lib/src/coin/dot/utils.ts:202]: What is the reason for this change that split into 2 functions?

File: modules/account-lib/src/coin/dot/transactionBuilder.ts [hunk 5]
```diff
@@ -194,9 +194,10 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
     if (utils.isSigningPayload(decodedTxn)) {
       this.referenceBlock(decodedTxn.blockHash);
     } else {
-      this.sender({ address: utils.decodeDotAddress(decodedTxn.address) });
+      const keypair = utils.decodeDotAddressToKeyPair(decodedTxn.address);
```
Comment (byhow) [inline modules/account-lib/src/coin/dot/transactionBuilder.ts:197]: ditto: reasons for changes

**Review by byhow — COMMENTED** [review]
some nits, and the tests seem to be failing

File: modules/account-lib/src/coin/dot/transactionBuilder.ts (hunk 5 above)
Comment (jrafael-simtlix) [reply modules/account-lib/src/coin/dot/transactionBuilder.ts:197]: BuildImplementation sign a transaction when signature is added in transactionBuilder. 
Two line below, it changes from directly push the signature in the transaction to addSignature in the builder.
AddSignature requieres the public key. This changes is to get address and public key in one step.

https://github.com/BitGo/BitGoJS/blob/bc58eda7da65815cd79521b103e5584e82d84298/modules/account-lib/src/coin/dot/transactionBuilder.ts#L224-L227

**Comment by jrafael-simtlix** [comment]
> some nits, and the tests seem to be failing
We know. Our goal is pass that tests. That is the improve we look for.
We rebase and changes so many thing that this PR loose it draft status. 
May be we need a new PR to avoid review until it's ready.

**Comment by nachofm** [comment]
image
All tests passing in my local machine.

**Review by argjv — APPROVED** [review]

File: modules/account-lib/test/unit/coin/dot/transactionBuilder/transactionBuilderFactory.ts [hunk 6]
```diff
@@ -17,7 +17,7 @@ describe('dot Transaction Builder Factory', function () {
   let factory: TransactionBuilderFactory;
 
   // TODO: BG-43197
-  xdescribe('parse generic builders', function () {
+  describe('parse generic builders', function () {
```
Comment (lcovar) [inline modules/account-lib/test/unit/coin/dot/transactionBuilder/transactionBuilderFactory.ts:20]: @nachofm youre going to have to skip this test again so you can get a green build.

Comment (nachofm) [reply modules/account-lib/test/unit/coin/dot/transactionBuilder/transactionBuilderFactory.ts]: Done, drone is passing OK now

**Review by lcovar — DISMISSED** [review]

File: modules/account-lib/test/unit/coin/dot/transactionBuilder/transactionBuilderFactory.ts [hunk 7]
```diff
@@ -16,7 +16,6 @@ describe('dot Transaction Builder Factory', function () {
   const sender2 = accounts.account3;
   let factory: TransactionBuilderFactory;
 
-  // TODO: BG-43197
```
[the remark this answers is not in the document]
Comment (nachofm) [reply modules/account-lib/test/unit/coin/dot/transactionBuilder/transactionBuilderFactory.ts:19]: Done in the following line

**Review by byhow — APPROVED** [review]
